src/Processors/Formats/Impl/AvroRowOutputFormat.cpp

```diff
@@ -401,6 +398,45 @@ AvroSerializer::SchemaWithSerializeFn AvroSerializer::createSchemaWithSerializeF
         }
         case TypeIndex::Nothing:
             return {avro::NullSchema(), [](const IColumn &, size_t, avro::Encoder & encoder) { encoder.encodeNull(); }};
+        case TypeIndex::Variant:
+        {
+            const auto & variant_type = assert_cast<const DataTypeVariant &>(*data_type);
+
+            avro::UnionSchema union_schema;
+            const auto & nested_types = variant_type.getVariants();
+
+            std::vector<SerializeFn> nested_serializers;
+            nested_serializers.reserve(nested_types.size());
+
+            for (const auto & nested_type : nested_types)
+            {
+                const auto [schema, serialize] = createSchemaWithSerializeFn(nested_type, type_name_increment, column_name);
+                union_schema.addType(schema);
+                nested_serializers.push_back(serialize);
+            }
+
+            // Since Variants have no schema-guaranteed nullability, we need to always include the null as one of the options in Avro Union.
+            // This is because Variant is considered Null in case it doesn't have any of the variants defined.
+            const auto nullUnionIndex = nested_types.size();
+            union_schema.addType(avro::NullSchema());
+
+            return {static_cast<avro::Schema>(union_schema), [serializers = std::move(nested_serializers), nullUnionIndex](const IColumn & column, const size_t row_num, avro::Encoder & encoder)
+            {
+                const auto & col = assert_cast<const ColumnVariant &>(column);
+                const auto global_discriminator = col.globalDiscriminatorAt(row_num);
+
+                if (global_discriminator == ColumnVariant::NULL_DISCRIMINATOR)
+                {
+                    encoder.encodeUnionIndex(nullUnionIndex);
+                    encoder.encodeNull();
+                }
+                else
+                {
+                    encoder.encodeUnionIndex(global_discriminator);
```

Comment:
It seems you're trying to adjust the position of an index in an Avro schema. However, the getVariants function appears to return internal types based on the order of local discriminators, not the global ones. I think you should be using the local discriminator here instead. Could you please review this piece of code one more time to make sure?

Reply:
`variant_type.getVariants()` returns variants in the global order. This code looks ok to me

Reply:
```
   const ColumnPtr & getVariantPtrByLocalDiscriminator(size_t discr) const { return variants[discr]; }
    const ColumnPtr & getVariantPtrByGlobalDiscriminator(size_t discr) const { return variants[global_to_local_discriminators.at(discr)]; }
    ColumnPtr & getVariantPtrByLocalDiscriminator(size_t discr) { return variants[discr]; }
    ColumnPtr & getVariantPtrByGlobalDiscriminator(size_t discr) { return variants[global_to_local_discriminators.at(discr)]; }

    const NestedColumns & getVariants() const { return variants; }
    NestedColumns & getVariants() { return variants; }
```
Are you sure? It seems to me that we index it directly by local discriminator, not global

Reply:
We use `DataTypeVariant::getVariants` not `ColumnVariant::getVariants` here

Reply:
Agree. Thanks for explaining